### ISSUE TITLE
Request via `globalThis.fetch` is not intercepted by MockAgent

### ISSUE THREAD

**Issue by mdrobny** [nodejs/undici#1882]
## Bug Description

I am facing a problem when intercepting and mocking responses from external APIs when using Node.js built-in `fetch`.

I am following [MockAgent docs](https://undici.nodejs.org/#/docs/api/MockAgent) and I noticed that it doesn't work when using [Jest](https://github.com/facebook/jest) as test runner.

Intercepting works correctly only when
- I am using `fetch` imported from `undici` 
- or I don't use Jest

Do you have any ideas what's causing the problem in Jest?
I investigated it a lot and currently I have no clue. Probably some internal knowledge about `undici` is needed.

## Reproducible By

`jest.test.js`

```js
const assert = require('node:assert');
const { MockAgent, setGlobalDispatcher, fetch: undiciFetch } = require('undici');

const mockAgent = new MockAgent();
mockAgent.disableNetConnect();
setGlobalDispatcher(mockAgent);

const apiUrl = 'http://test-api.com';

test('Test', async () => {
    const mockPool = mockAgent.get(apiUrl);
    mockPool.intercept({ method: 'GET', path: '/foo' }).reply(200, { foo: 'bar' }).persist()

    // Works!
    const undiciFetchResponse = await undiciFetch(`${apiUrl}/foo`);
    assert.equal(undiciFetchResponse.status, 200); // ok

    // Doesn't work - "TypeError: fetch failed"
    const fetchResponse = await globalThis.fetch(`${apiUrl}/foo`);
    assert.deepEqual(fetchResponse.status, 200);

    // Works!
    const fetchWithDispatcherResponse = await globalThis.fetch(`${apiUrl}/foo`, { dispatcher: mockAgent });
    assert.deepEqual(fetchWithDispatcherResponse.status, 200);
});
```

Executing
```shell
npx jest jest.test.js
```
returns error

```shell
 FAIL  ./jest.test.js
  ✕ Test (288 ms)

  ● Test

    TypeError: fetch failed

      17 |
      18 |     // Doesn't work - "TypeError: fetch failed"
    > 19 |     const fetchResponse = await globalThis.fetch(`${apiUrl}/foo`);
         |                           ^
      20 |     assert.deepEqual(fetchResponse.status, 200); // fails
      21 |

      at Object.<anonymous> (jest.test.js:19:27)
```

It fails the same when using Mocha runner.

## Expected Behavior

`node.test.mjs`

```js
import test from 'node:test';
import assert from 'node:assert';
import { MockAgent, setGlobalDispatcher } from 'undici';

const mockAgent = new MockAgent();
mockAgent.disableNetConnect();
setGlobalDispatcher(mockAgent);

const apiUrl = 'http://test-api.com';

test('Test', async (t) => {
    const mockPool = mockAgent.get(apiUrl);
    mockPool.intercept({ method: 'GET', path: '/foo' }).reply(200, { foo: 'bar' }).persist()

    // Works!
    const fetchResponse = await globalThis.fetch(`${apiUrl}/foo`);
    assert.deepEqual(fetchResponse.status, 200); // ok
    assert.deepEqual(await fetchResponse.json(), { foo: 'bar' }); // ok
});
```

Executing
```shell
node node.test.mjs
```

```shell
TAP version 13
# Subtest: Test
ok 1 - Test
  ---
  duration_ms: 9.547417
  ...
1..1
# tests 1
# pass 1
# fail 0
# cancelled 0
# skipped 0
# todo 0
# duration_ms 15.368792
```

## Environment

- Node.js - 18.13.0
- undici - 5.15.0
- Jest - 29.3.1
- MacOS - 12.6.2

### Additional context

I am replacing `node-fetch` in my application with Node.js built-in `fetch` so I have to replace [nock](https://github.com/nock/nock) with `undici` in tests for mocking API responses but I am stuck with the migration because of this problem.

**Comment by KhafraDev** [comment]
To fix the third case:
```diff
-     mockPool.intercept({ method: 'GET', path: '/foo' }).reply(200, { foo: 'bar' })
+     mockPool.intercept({ method: 'GET', path: '/foo' }).reply(200, { foo: 'bar' }).persist()
```

The second one is probably an issue with jest. Not sure if we can debug it without having the error's cause.

**Comment by mdrobny** [comment]
@KhafraDev you are correct, adding `.persist()` fixes the test with `fetchWithDispatcherResponse` 👍🏼 
I will fix code snippets and description.

> The second one is probably an issue with jest

That's true also 👍🏼 It works correctly in Mocha after adding `.persist()`. 

So indeed it sounds like there is a problem with Jest so this issue rather belongs to [Jest](https://github.com/facebook/jest) repository. But I guess many developers may face this problem in the future since Jest is popular so it would be nice to get some help in solving it ;) 

> Not sure if we can debug it without having the error's cause.

I am not sure how to find error's cause yet. `fetch` function just rejects

**Comment by mcollina** [comment]
Maybe I don't understand this issue but the test in the expected "Expected Behavior" is passing. I think it's something related to jest. I don't know why but the global symbols are not the same, but it looks another instance of https://github.com/facebook/jest/issues/2549:


```js
const assert = require('node:assert');
const { MockAgent, setGlobalDispatcher, fetch: undiciFetch } = require('undici');

const apiUrl = 'http://test-api.com';
const mockAgent = new MockAgent();
mockAgent.disableNetConnect();
setGlobalDispatcher(mockAgent);

test('Test', async () => {

    const mockPool = mockAgent.get(apiUrl);
    mockPool.intercept({ method: 'GET', path: '/foo' }).reply(200, { foo: 'bar' }).persist();

    // Works!
    const undiciFetchResponse = await undiciFetch(`${apiUrl}/foo`);
    assert.equal(undiciFetchResponse.status, 200); // ok

    // works!
    const fetchWithDispatcherResponse = await globalThis.fetch(`${apiUrl}/foo`, { dispatcher: mockAgent });
    assert.deepEqual(fetchWithDispatcherResponse.status, 200);

    // "TypeError: fetch failed"
    const fetchResponse = await globalThis.fetch(`${apiUrl}/foo`);
    assert.deepEqual(fetchResponse.status, 200);
});
```

**Comment by KhafraDev** [comment]
That's a good point, I assumed the global symbol would work between vm contexts but that's definitely the issue. There's nothing we can do if that's the case so I'll close this.

**Comment by nickcampbell18** [comment]
I spent a day fighting this today at work and came up with an alternative solution that I would like to share. My goals were:

* Able to replace the MockAgent for each test (so interceptor state doesn't persist between different tests)
* I can silently intercept unmodified `fetch` calls in both my code and library code
* Solution works in both Jest single-process (e.g. `maxWorkers: 1`) and multi-process (i.e. child processes or worker threads)

I wrote it up as a little gist: https://gist.github.com/nickcampbell18/5ed0539929fb8dbc8a12c974c7b00e1f.

Essentially, because Jest is messing around with `global` _anyway_, we just replace `global.fetch` with a wrapper that uses the correct dispatcher that we control. And it doesn't use Symbols, which seem to vary between Jest contexts, but just a plain string to index the global object with our custom dispatcher. I added a special global set method with a special name so you don't confuse it with `setGlobalDispatcher` from this library.
